core/src/test/scala/integration/kafka/server/RaftClusterTest.scala

```diff
@@ -313,4 +318,94 @@ class RaftClusterTest {
       cluster.close()
     }
   }
+
+  @Test
+  def testCreateClusterWithAdvertisedPortZero(): Unit = {
+    val nodes = new TestKitNodes.Builder()
+      .setNumControllerNodes(1)
+      .setNumBrokerNodes(3)
+      .build()
+    nodes.brokerNodes().values().forEach { broker =>
+      broker.propertyOverrides().put(KafkaConfig.ListenersProp,
+        s"${nodes.externalListenerName().value()}://localhost:0")
+      broker.propertyOverrides().put(KafkaConfig.AdvertisedListenersProp,
+        s"${nodes.externalListenerName().value()}://localhost:0")
+    }
+    val cluster = new KafkaClusterTestKit.Builder(nodes).build()
+    try {
+      cluster.format()
+      cluster.startup()
+
+      val (runningBrokerServer, foundRunningBroker) = TestUtils.computeUntilTrue(anyBrokerServer(cluster)){
+        brokerServer => brokerServer.currentState() == BrokerState.RUNNING
+      }
+      assertTrue(foundRunningBroker, "No Broker never made it to RUNNING state.")
+
+      val (describeClusterResponse, metadataUpToDate) = TestUtils.computeUntilTrue(
+        sendDescribeClusterRequest(runningBrokerServer.socketServer, nodes.externalListenerName())
+      ) {
+        response => response.nodes().size() == nodes.brokerNodes().size()
+      }
+      assertTrue(metadataUpToDate, s"Broker never reached expected cluster size of ${nodes.brokerNodes().size()}")
```

Comment:
nit: these tests are basically the same except for the initialization and assertions. It would be nice to factor out a helper or two.

Reply:
I refactored the two tests and also considered your other two remarks.